### PR TITLE
Cleanup basic shape code

### PR DESCRIPTION
**Pull request by ghost**
I began with isolating serialization stuff for `BorderRadius` to help #16340 and #16352, but ended up cleaning a lot of code in `basic_shape.rs` (it's just a cleanup, should be easy to review)

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [x] These changes do not require tests because it's a cleanup

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16383)

### CODE REVIEW

File: components/style/values/specified/basic_shape.rs [hunk 1]
```diff
@@ -465,28 +450,18 @@ pub struct Ellipse {
     pub position: Position,
 }
 
-
 impl Ellipse {
     #[allow(missing_docs)]
     pub fn parse_function_arguments(context: &ParserContext, input: &mut Parser) -> Result<Ellipse, ()> {
-        let (a, b) = input.try(|input| -> Result<_, ()> {
-            Ok((try!(ShapeRadius::parse(context, input)), try!(ShapeRadius::parse(context, input))))
+        let (a, b) = input.try(|i| -> Result<_, ()> {
```
Comment (emilio) [inline components/style/values/specified/basic_shape.rs]: This type annotation in the closure should not be needed, right?

**Review by emilio — COMMENTED** [review]
Nice cleanup! r=me with that :)

File: components/style/values/specified/basic_shape.rs (hunk 1 above)
Comment (Wafflespeanut) [inline components/style/values/specified/basic_shape.rs]: Thought the same, but it looks like `rustc` cannot infer the type of error if I remove it. That's probably the reason why it exists in the first place.

    error[E0282]: unable to infer enough type information about `E`
       --> /media/Ubuntu/servo/components/style/values/specified/basic_shape.rs:456:28
        |
    456 |         let (a, b) = input.try(|i| {
        |                            ^^^ cannot infer type for `E`
        |
        = note: type annotations or generic parameter binding required

I could use `and_then`, but that might get a bit nasty.

**Comment by Wafflespeanut** [comment]
Thanks for the review! :)

@bors-servo r=emilio

**Comment by jdm** [comment]
From `./mach build-geckolib`:
```
   Compiling style v0.0.1 (file:///Users/servo/buildbot/slave/mac-dev-unit/build/components/style)
error[E0277]: the trait bound `values::specified::basic_shape::ShapeBox: parser::Parse` is not satisfied
     --> /Users/servo/buildbot/slave/mac-dev-unit/build/target/geckolib/debug/build/style-12d7824bce835f13/out/properties.rs:18985:9
      |
18985 |         ShapeSource::parse(context, input)
      |         ^^^^^^^^^^^^^^^^^^ the trait `parser::Parse` is not implemented for `values::specified::basic_shape::ShapeBox`
      |
      = note: required because of the requirements on the impl of `parser::Parse` for `values::specified::basic_shape::ShapeSource<values::specified::basic_shape::ShapeBox>`
      = note: required by `parser::Parse::parse`

error: aborting due to previous error

error: Could not compile `style`.
```

**Comment by Wafflespeanut** [comment]
@emilio #16384 uses `ShapeSource<ShapeBox>`, but since `ShapeBox` is defined using `define_css_keyword_enum!`, it doesn't `impl Parse`, and moving the macro to `style` crate or moving the trait to `style_traits` may demand more work. So, I've gone for another macro for implementing the necessary traits to macro-defined keyword enums. You okay with this?

File: components/style/values/mod.rs [hunk 2]
```diff
@@ -60,6 +60,23 @@ macro_rules! no_viewport_percentage {
     };
 }
 
+/// A macro for implementing `ComputedValueAsSpecified`, `Parse`
+/// and `HasViewportPercentage` traits for the enums defined
+/// using `define_css_keyword_enum` macro.
```
Comment (emilio) [inline components/style/values/mod.rs]: Please reference that a proper solution would be to move `Parse` to style_traits, but that may not be trivial.

**Review by emilio — COMMENTED** [review]
Looks reasonable, but please add those comments.

**Comment by Wafflespeanut** [comment]
Added those comments. Thanks! :)

@bors-servo r=emilio

**Pull request merged**